tracing-flame/src/lib.rs

```diff
@@ -0,0 +1,396 @@
+use std::io::Write;
+use std::marker::PhantomData;
+use std::path::Path;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::time::{Duration, Instant};
+use tracing::span::Attributes;
+use tracing::Id;
+use tracing::Subscriber;
+use tracing_subscriber::layer::Context;
+use tracing_subscriber::registry::LookupSpan;
+use tracing_subscriber::registry::SpanRef;
+use tracing_subscriber::Layer;
+
+mod error;
+
+lazy_static! {
+    static ref START: Instant = Instant::now();
+}
+
+thread_local! {
+    static LAST_EVENT: Cell<Instant> =  Cell::new(*START);
+}
+
+/// A `Layer` that records span open/close events as folded flamegraph stack
+/// samples.
+///
+/// The output of `FlameLayer` emulates the output of commands like `perf` once
+/// they've been collapsed by `inferno-flamegraph`. The output of this layer
+/// should look similar to the output of the following commands:
+///
+/// ```sh
+/// perf record --call-graph dwarf target/release/mybin
+/// perf script | inferno-collapse-perf > stacks.folded
+/// ```
+///
+/// # Sample Counts
+///
+/// Because `tracing-flame` doesn't use sampling, the number at the end of each
+/// folded stack trace does not represent a number of samples of that stack.
+/// Instead, the numbers on each line are the number of nanoseconds since the
+/// last event in the same thread.
+///
+/// # Dropping and Flushing
+///
+/// If you use a global subscriber the drop implementations on your various
+/// layers will not get called when your program exits. This means that if
+/// you're using a buffered writer as the inner writer for the `FlameLayer`
+/// you're not guaranteed to see all the events that have been emitted in the
+/// file by default.
+///
+/// To ensure all data is flushed when the program exits, `FlameLayer` exposes
+/// the [`flush_on_drop`] function, which returns a [`FlushGuard`]. The `FlushGuard`
+/// will flush the writer when it is dropped. If necessary, it can also be used to manually
+/// flush the writer.
+///
+/// [`flush_on_drop`]: struct.FlameLayer.html#method.flush_on_drop
+/// [`FlushGuard`]: struct.FlushGuard.html
+#[derive(Debug)]
+pub struct FlameLayer<S, W> {
+    out: Arc<Mutex<W>>,
+    _inner: PhantomData<S>,
+}
+
+/// An RAII guard for managing flushing a global writer that is
+/// otherwise inaccessible.
+///
+/// This type is only needed when using
+/// `tracing::subscriber::set_global_default`, which prevents the drop
+/// implementation of layers from running when the program exits.
+#[must_use]
+#[derive(Debug)]
+pub struct FlushGuard<W>
+where
+    W: Write + 'static,
+{
+    out: Arc<Mutex<W>>,
+}
+
+impl<S, W> FlameLayer<S, W>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+    W: Write + 'static,
+{
+    /// Returns a new `FlameLayer` that outputs all folded stack samples to the
+    /// provided writer.
+    pub fn new(writer: W) -> Self {
+        // Initialize the start used by all threads when initializing the
+        // LAST_EVENT when constructing the layer
```

Comment:
I think adding some code that references `START` would fix the panic I mentioned in a separate comment. We could just do:
```suggestion
        // Initialize the start used by all threads when initializing the
        // LAST_EVENT when constructing the layer
        let _unused = *START;
```
or something to force it to be initialized.